File: RinhabackendGo/repository/ClientRepository.go
package repository

import (
	"RinhabackendGo/model"
	"context"
	"errors"
	"github.com/gocraft/dbr/v2"
)

type ClientRepository struct {
	DbConnection *dbr.Connection
}

func (c ClientRepository) MakeTransaction(ctx context.Context, transaction model.Transaction) (model.Client, error) {

	session := c.DbConnection.NewSession(nil)
	tx, _ := session.BeginTx(ctx, nil)
	defer tx.Commit()

	balanceStatement := tx.SelectBySql(`
	SELECT valor, limite FROM clientes WHERE id = ? LIMIT 1 FOR NO KEY UPDATE 
	`, transaction.ClientId)
	client := model.Client{}
	err := balanceStatement.LoadOne(&client)
	if err != nil {
		tx.Rollback()
		return model.Client{}, err
	}
	transactionValue := transaction.Value
	if transaction.Type == "d" {
		transactionValue = -transactionValue
	}
	client.Value = client.Value + transactionValue
	if client.Value < -client.Limit {
		tx.Rollback()
		return model.Client{}, errors.New("saldo insuficiente")
	}

	updateStatement := tx.UpdateBySql(`
 		UPDATE clientes SET valor = ? WHERE id = ?
 	`, client.Value, transaction.ClientId)

	_, err = updateStatement.Exec()
	if err != nil {
		tx.Rollback()
		return model.Client{}, err
	}
	transationStatement := tx.InsertBySql(`
		INSERT INTO transacoes(id, valor, cliente_id, tipo, descricao, realizada_em)
		    			values (default, ?, ?, ?, ?, now())
		`, transaction.Value, transaction.ClientId, transaction.Type, transaction.Description)

	_, err = transationStatement.Exec()
	if err != nil {
		tx.Rollback()
		return model.Client{}, err
	}
	return client, nil
}

//func (c ClientRepository) MakeTransactionSQL(ctx context.Context, transaction model.Transaction) (model.Client, error) {
//
//	session := c.DbConnection.NewSession(nil)
//	tx, _ := session.BeginTx(ctx, nil)
//	defer tx.Commit()
//	transactionValue := transaction.Value
//	if transaction.Type == "d" {
//		transactionValue = -transactionValue
//	}
//
//	balanceStatement := tx.SelectBySql(`
//		SELECT * FROM make_transaction(?, ?, ?, ?)
//	`, transaction.ClientId, transactionValue, transaction.Type, transaction.Description)
//	client := model.Client{}
//	err := balanceStatement.LoadOne(&client)
//	if err != nil {
//		tx.Rollback()
//		return model.Client{}, err
//	}
//	return client, nil
//}

func (c ClientRepository) GetExtract(ctx context.Context, id int) (model.Extract, error) {
	session := c.DbConnection.NewSession(nil)
	clientStatement := session.SelectBySql(`
	SELECT limite, valor, now() as data_extrato FROM clientes where id = ?
	`, id)

	var client model.ClientBalance
	err := clientStatement.LoadOne(&client)

	if err != nil {
		if errors.Is(err, dbr.ErrNotFound) {
			return model.Extract{}, errors.New("Cliente não encontrado")
		}
		return model.Extract{}, err
	}

	transactionsStatement := session.SelectBySql(`
	SELECT valor, tipo, descricao, realizada_em FROM transacoes where cliente_id = ? ORDER BY realizada_em DESC LIMIT 10
	`, id)

	var transactions []model.TransactionBalance
	_, err = transactionsStatement.Load(&transactions)
	if err != nil {
		return model.Extract{}, err
	}

	extract := model.Extract{
		Balance:     client,
		Transaction: transactions,
	}

	return extract, nil
}

File: RinhaBackendSpringBoot/src/main/java/br/com/ronistone/rinhabackendspringboot/repository/ClientRepository.java
package br.com.ronistone.rinhabackendspringboot.repository;

import br.com.ronistone.rinhabackendspringboot.model.ClientBalance;
import br.com.ronistone.rinhabackendspringboot.model.Extract;
import br.com.ronistone.rinhabackendspringboot.model.Transaction;
import br.com.ronistone.rinhabackendspringboot.model.TransactionBalance;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.OffsetDateTime;

@Component
public class ClientRepository {

    private JdbcTemplate jdbcTemplate;

    public ClientRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    @Transactional
    public ClientBalance makeTransaction(Transaction transaction) {
        var clientResul = jdbcTemplate.query("SELECT valor, limite FROM clientes WHERE id = ? LIMIT 1 FOR NO KEY UPDATE", (rs, rowNum) -> {
            Integer value = rs.getInt("valor");
            Integer limit = rs.getInt("limite");
            if (transaction.type.equals("c")) {
                value += transaction.value;
            } else {
                if (value - transaction.value < -limit) {
                    throw new RuntimeException("Insufficient funds");
                }
                value -= transaction.value;
            }
            jdbcTemplate.update("UPDATE clientes SET valor = ? WHERE id = ?", value, transaction.clientId);
            jdbcTemplate.update("INSERT INTO transacoes (cliente_id, valor, tipo, descricao, realizada_em) VALUES (?, ?, ?, ?, now())", transaction.clientId, transaction.value, transaction.type, transaction.description);
            return new ClientBalance(limit, value, transaction.createAt);
        }, transaction.clientId);

        if (clientResul.isEmpty()) {
            throw new RuntimeException("Client not found");
        }
        return clientResul.getFirst();
    }

    public Extract getExtract(Integer clientId) {
        Extract extract = new Extract();
        jdbcTemplate.query("SELECT limite, valor, now() as data_extrato FROM clientes where id = ?", (rs, rowNum) -> {
            ClientBalance clientBalance = new ClientBalance(rs.getInt("limite"), rs.getInt("valor"), rs.getObject("data_extrato", OffsetDateTime.class));
            extract.balance = clientBalance;

            jdbcTemplate.query("SELECT valor, tipo, descricao, realizada_em FROM transacoes where cliente_id = ? ORDER BY realizada_em DESC LIMIT 10", (rs2, rowNum2) -> {
                if(extract.transactions == null) {
                    extract.transactions = new java.util.ArrayList<>();
                }
                extract.transactions.add(new TransactionBalance(rs2.getInt("valor"), rs2.getString("tipo"), rs2.getString("descricao"), rs2.getObject("realizada_em", OffsetDateTime.class)));
                return null;
            }, clientId);
            return null;
        }, clientId);

        return extract;
    }

    public void reset() {
        jdbcTemplate.update("DELETE FROM transacoes");
        jdbcTemplate.update("UPDATE clientes SET valor = 0");
    }

    public void checkHealth() {
        jdbcTemplate.query("SELECT 1", (rs, rowNum) -> null);
    }


}
